Migrate reference documentation to Antora

pderop opened the pull request:
This PR attempts to migrate Reactor Core reference documentation to Antora, and includes the following logical commits:

- a482fbbd007eb2f96fe1748e355acb40ae5f6f63: Moved asciidoc file paths so they are compliant with Antora structure:
- 2d340c040cb21309c3174d115a320d8d244fd6f6: Refactor ascidoc adoc files to Antora. core features and advenced features are now split in multiple adoc files
- 90e86ee64cc884bb0aec5ca71515cea1b3c0f61d: Add antora yml file descritors
- 33be68a3b15becb7598a61a80dde79d361c9eca0: Adapted gradle scripts for Antora
- eaa7bb44303843d8e1f25d06ed69c88037b004cb: Adapted the `publish` CI in order to configure JDK21 as the default JDK, because Antora gradle plugin requires a JDK17 compatible version (so far, I could not use toolchain to selectively run only the antora gradle plugin with JDK21, because Gradle loads all plugin classes in the current JVM used, see this [discussion](https://stackoverflow.com/questions/75625330/gradle-toolchain-resolver-plugin-doesnt-download-jdk17-for-a-spring-boot-3-proj) which might be related). Setting JDK21 as the default JDK should not affect generation of all jars because toolchains still ensures that jars are built with proper JDK version.  Note that if using JDK21 as the default JDK is really an issue, we can consider to do what was done for reactor-netty: use a separate job that builds only the doc in JDK21, then share the docs.zip between jobs using some upload/download GH actions ...). In the commit, the asciidoctor-pdf tool is also installed when releasing a RELEASE , or a MILESTONE version, because this tool is needed when converting antora to PDF (in snapshot mode, PDF is not generated, so asciidoctor-pdf tool is not installed in this case).
- acec348b730fb2df5030a32767b65ff560843b6c: Updated README to explain how to generate the documentation locally.

### To test this PR:

- Set the current JDK to JDK21 (or a JDK17 compatible version)
- generate the documentation:
```
./gradlew antora
``` 
you can browse docs/build/site/index.html

The search is not yet enabled, it will be enabled once we'll have merged this PR because it's needed to crawl the documentation using Algolia from **projectreactor.io** site.

### To test docs.zip generation:

- Set the current JDK to JDK21 (or a JDK17 compatible version)
- then type:
```
./gradlew docs
```

it will generate docs/build/distributions/reactor-core-3.7.0-SNAPSHOT-docs.zip

### To test publication:

- make sure all JDK versions are properly installed (see `Trouble building the project` section in the reactor core [README](https://github.com/reactor/reactor-core)).
- set current JDK version to JDK21.
- Run publish command:
```
./gradlew publishToMavenLocal
```
All jars should be built with proper jdk versions because toolchain is used.
- then check what is generated in your M2 repo

### To test publication in two steps:

If using JDK21 as the default JDK version is really an issue, you can then publish everything in two steps:

- step 1: Set current JDK to JDK21, and only run `./gradlew docs` : this will only build the docs.zip
- step 2: Set current JDK to JDK8, then run `./gradlew publishToMavenLocal`, this command will pick up the docs.zip previously built in step 1 and will include it in local M2

### To test PDF generation

As before, PDF are not generated by default in snapshot mode. If you want to generate a PDF, then do the following:

- first, you will need to install `asciidoctor-pdf` command, because the antora pdf assembler requires it:
```
brew install asciidoctor
```

- then generate antora documention + PDF:

```
./gradlew antora -PforcePdf
```

docs is generated as before in docs/build/site/index.html, but this time a PDF is also generated in ./docs/build/assembler/reactor/3.7.0/reactor-3-reference-guide.pdf

If you want to include the pdf in the docs.zip distribution file:

```
./gradlew docs -PforcePdf
```

and the pdf will be generated and included in the docs.zip:
```
unzip -l docs/build/distributions/reactor-core-3.7.0-SNAPSHOT-docs.zip |grep pdf
  4114787  05-03-2024 10:35   docs/reactor-core-reference-guide-3.7.0-SNAPSHOT.pdf
```

- you can force PDF generation also when publishing:
```
./gradlew publishToMavenLocal -PforcePdf
unzip -l ~/.m2/repository/io/projectreactor/reactor-core/3.7.0-SNAPSHOT/reactor-core-3.7.0-SNAPSHOT-docs.zip | grep pdf
  4114787  05-03-2024 10:37   docs/reactor-core-reference-guide-3.7.0-SNAPSHOT.pdf
```

as before, PDF is generated and included when releasing a RELEASE, or RC, or a MILESTONE, but not when releasing a SNAPSHOT version.

Fixes #3083

pderop commented:
the previous force-push commit contains a fix in the GH action publish.yml, in order to build the docs-zip using a separate build-docs-zip job that is using JDK17. This allows the other deploy jobs (deployRelease/deploySnapshot/deployMilistone) to use JDK8 as the default JDK.

chemicL commented on .github/workflows/publish.yml:
```diff
@@ -63,11 +68,41 @@ jobs:
+
+  # Build the docs-zip antora doc for the current branch, and upload generated docs-zip to workflow run.
+  # JDK21 is used because the antora plugin requires a JDK17 compatible version.
+  # Each deploy jobs can then download the docs-zip into ./docs/build/distributions/ in order to let it be included in published artifacts.
```
```suggestion
  # Each deploy job can then download the docs-zip to ./docs/build/distributions/ in order to be included in published artifacts.
```

pderop replied:
fixed in https://github.com/reactor/reactor-core/pull/3806/commits/dfe2a219a32b03874e620f1d97f879acef81d969

chemicL commented on README.md:
```diff
@@ -89,6 +89,18 @@ When the installations succeed, try to refresh the project and see that it build
+### Building the doc
+
+The current active shell JDK version must be compatible with JDK17 or higher for Antora to build successfully.
+So, just ensure that you have installed JDK 21, as previously described and make it as the current one. 
```
```suggestion
So, just ensure that you have installed JDK 21, as described above and make it as the current one. 
```

pderop replied:
fixed in https://github.com/reactor/reactor-core/pull/3806/commits/8aff270d7287936d0607136c0ca2c12a8c21fff2

chemicL commented on gradle/libs.versions.toml:
```diff
@@ -12,14 +12,22 @@ baselinePerfCore = "3.6.5"
 contextPropagation="1.1.1"
 kotlin = "1.8.22"
 reactiveStreams = "1.0.4"
+antora = "1.0.0"
+antora-yml = "0.0.1"
+antora-version = "3.2.0-alpha.4"
+antora-atlas-extension = "1.0.0-alpha.1"
+antora-pdf-extension = "1.0.0-alpha.7"
+antora-collector-extension = "1.0.0-alpha.3"
+antora-tabs-extension = "1.0.0-beta.6"
+antora-springio-antora-extension = "1.8.2"
+antora-asciidoctor-extension = "1.0.0-alpha.9"
```
Why are these versions not unified?

chemicL commented on gradle/setup.gradle:
```diff
@@ -112,10 +112,28 @@ publishing {
+				if (docsProject) {
+					// Access the docsZip task from the docs project
+					def docsZipTask = docsProject.tasks.findByName('docsZip')
+					if (docsZipTask) {
```
Should there not be a failure if this task is not defined?

pderop replied:
first, there was a bug in this part: the docs zip should be included only in the `reactor-core` artifact, not in all artifacts. So, I added the missing check in the `afterEvaluate` section:

```
if (project.name == 'reactor-core') {
```
see https://github.com/reactor/reactor-core/pull/3806/commits/34c854e6ff3b2234a3bb57e858e33b24be719b36

Also, did this in this commit:
- if the current JDK version is greater than JDK8, then include the docsZip artifact as it was done previously, there is no need to change what was done before:
```
artifact docsProject.tasks.docsZip
```
- if the current JDK version is lower than JDK8, and if no docs zip archive is found from the docs/build/distributions/ directory, then add a WARN message to indicate to the user that the docs won't be published:
```
logger.warn("Antora docs zip not found from docs/ project. Use JDK 17+ to include docs in published artifacts.")
```
- finally, fixed and clarified comments

chemicL commented on gradle/setup.gradle:
```diff
@@ -112,10 +112,28 @@ publishing {
 			afterEvaluate {
-				if (project.name == 'reactor-core') {
-					artifact rootProject.tasks.docsZip
+				// Find the docs project (it's not a java project, so we need to lookup it)
```
```suggestion
				// Find the docs project (it's not a java project, so we need to look it up)
```

pderop replied:
fixed and clarified comments, se ehttps://github.com/reactor/reactor-core/commit/34c854e6ff3b2234a3bb57e858e33b24be719b36

chemicL commented on gradle/setup.gradle:
```diff
@@ -112,10 +112,28 @@ publishing {
+					}
+				}
+				else {
+					// If the current JDK version is JDK8, the docs project is not loaded (see settings.gradle), so
```
```suggestion
					// If the current JDK version is lower than JDK17, the docs project is not loaded (see settings.gradle), so
```

pderop replied:
fixed and clarified comments, se ehttps://github.com/reactor/reactor-core/commit/34c854e6ff3b2234a3bb57e858e33b24be719b36

chemicL commented on gradle/setup.gradle:
```diff
@@ -112,10 +112,28 @@ publishing {
+				}
+				else {
+					// If the current JDK version is JDK8, the docs project is not loaded (see settings.gradle), so
+					// docsTask is not available. In this case, include the docsZip file path directly as an artifact, if it exists.
```
```suggestion
					// docsTask is not available. In this case, include the docsZip file path directly as an artifact, if it exists
```

pderop replied:
fixed and clarified comments, se ehttps://github.com/reactor/reactor-core/commit/34c854e6ff3b2234a3bb57e858e33b24be719b36

chemicL commented on gradle/setup.gradle:
```diff
@@ -112,10 +112,28 @@ publishing {
+					// docsTask is not available. In this case, include the docsZip file path directly as an artifact, if it exists.
+					// (it may exist in case the docs have been previously built using a JDK17 compatible JDK).
+					def docsZipFile = file("${rootDir}/docs/build/distributions/reactor-core-${project.version}-docs.zip")
+					if (docsZipFile.exists()) {
```
If it's missing it's still a desired outcome? Should we print a log message whether it is skipped?

pderop replied:
Added a WARN message, fixed and clarified comments, see ehttps://github.com/reactor/reactor-core/commit/34c854e6ff3b2234a3bb57e858e33b24be719b36

pderop commented on README.md:
````diff
@@ -89,6 +89,18 @@ When the installations succeed, try to refresh the project and see that it build
+./gradlew docs
+```
+
+The documentation is generated in `docs/build/site/index.html` and in `distribution docs/build/distributions/reactor-core-<version-docs.zip` 
````
I was meaning about the distributable artifacts generated by Gradle during the build process in the docs/build/distributions/ directory.

ok, that's confusing, I will just remove the 'distribution' prefix, like:

_The documentation is generated in `docs/build/site/index.html` and in `docs/build/distributions/reactor-core-<version-docs.zip`_

pderop replied:
fixed in https://github.com/reactor/reactor-core/pull/3806/commits/c6641fef6c6c991e6fbbc83514da9d619f8cda7e

pderop commented on docs/antora-assembler.yml:
```diff
@@ -0,0 +1,12 @@
+root_level: 0
+#component_versions: '*'
```
The PDF converter can generate multiple PDF and I used `component_versions` attribute to test the abilty to select individual versions, but I simply forgot to remove this attribute , we do not need it.

I have removed this commented attribute in https://github.com/reactor/reactor-core/pull/3806/commits/c1cefd826b3dd28b7f0e3479a34bf3f6d4c13e25

pderop commented on docs/antora-assembler.yml:
```diff
@@ -0,0 +1,12 @@
+    attributes:
+        source-highlighter: rouge
+        pdf-theme: ./pdf-theme.yml
+
```
Removed empty lines in https://github.com/reactor/reactor-core/pull/3806/commits/3f142067bc15097229f0bf6b15950e45a3a10bd7

pderop commented on docs/antora-playbook.yml:
```diff
@@ -0,0 +1,52 @@
+    hide-uri-scheme: '@'
+    tabs-sync-option: '@'
+    chomp: 'all'
+    # Adhust the following attributes (they comes from reactor-netty)
```
This comment was temporary, and should have been removed.
I just have removed it from https://github.com/reactor/reactor-core/pull/3806/commits/05ebe58c1756fd31520948de04657e196315d580

pderop commented on docs/build.gradle:
```diff
@@ -0,0 +1,214 @@
+}
+
+antora {
+    //version = "$antoraVersion"
```
Removed this useless comment in https://github.com/reactor/reactor-core/pull/3806/commits/5c8b21e4100260b9c0496918fa0aa699b281a60f

pderop commented:
in the last commit https://github.com/reactor/reactor-core/pull/3806/commits/20f7d80c1a16bf81b5452cdbeb68fff2c2f54d28, added docs/build/** in the nohttp section from the main build.gradle, else the "./gradlew checkstyleNohttp" fails.

chemicL commented:
Thanks for the updates. I'll check the generated docs now. First thing that happened was me running with JDK8:

    ./gradlew clean antora

```
FAILURE: Build failed with an exception.

* What went wrong:
Task 'antora' not found in root project 'reactor' and its subprojects.
```

I believe it should fail saying that JDK17+ is required instead.

chemicL commented:
Running

`./gradlew clean docs`

I get

```
FAILURE: Build failed with an exception.

* What went wrong:
Task 'docs' not found in root project 'reactor' and its subprojects.
```

I think it should also fail with a message explaining JDK17+ is required.

chemicL commented:
Keep in mind that projectreactor.io has a link to a specific section as "[Choosing an Operator](https://projectreactor.io/docs/core/release/reference/docs/index.html#which-operator)" that points to `index.html#which-operator`. After these changes it should point to `apdx-operatorChoice.html`.

chemicL commented:
This care should also be applied for the testing module that has a link on the site.

chemicL commented on docs/modules/ROOT/pages/aboutDoc.adoc:
```diff
@@ -26,12 +31,13 @@ rendering. Some of the sections rely on included files, so GitHub rendering is
-link:https://github.com/reactor/reactor-core/edit/main/docs/asciidoc/aboutDoc.adoc[Suggest Edit^, role="fa fa-edit"] to <<about-doc>>.
+TIP: To facilitate documentation edits, you can edit the current page from the `Edit this Page` link located in the upper right corner sidebar. The link opens
+an edit `UI` directly on `GitHub` for the main source file for the current page. These links are
+only present in the `HTML5` version of this reference guide. They look like the following link:
```
```suggestion
only present in the `HTML5` version of this reference guide. They look like the following:
```

chemicL commented on docs/modules/ROOT/pages/aboutDoc.adoc:
```diff
@@ -26,12 +31,13 @@ rendering. Some of the sections rely on included files, so GitHub rendering is
+TIP: To facilitate documentation edits, you can edit the current page from the `Edit this Page` link located in the upper right corner sidebar. The link opens
+an edit `UI` directly on `GitHub` for the main source file for the current page. These links are
+only present in the `HTML5` version of this reference guide. They look like the following link:
+link:https://github.com/reactor/reactor-core/edit/main/docs/modules/ROOT/pages/aboutDoc.adoc[Edit this Page^, role="fa fa-edit"] to xref:aboutDoc.adoc[About the Documentation].
```
```suggestion
link:https://github.com/reactor/reactor-core/edit/main/docs/modules/ROOT/pages/aboutDoc.adoc[Suggest Edit^, role="fa fa-edit"] to xref:aboutDoc.adoc[About the Documentation].
```

chemicL commented on docs/modules/ROOT/pages/aboutDoc.adoc:
```diff
@@ -26,12 +31,13 @@ rendering. Some of the sections rely on included files, so GitHub rendering is
-an edit UI directly on GitHub for the main source file for that section. These links are
-only present in the HTML5 version of this reference guide. They look like the following:
-link:https://github.com/reactor/reactor-core/edit/main/docs/asciidoc/aboutDoc.adoc[Suggest Edit^, role="fa fa-edit"] to <<about-doc>>.
+TIP: To facilitate documentation edits, you can edit the current page from the `Edit this Page` link located in the upper right corner sidebar. The link opens
```
```suggestion
TIP: To facilitate documentation edits, you can edit the current page from the `Suggest Edit` link located in the upper right corner sidebar. The link opens
```

chemicL commented on docs/modules/ROOT/pages/aboutDoc.adoc:
```diff
@@ -26,12 +31,13 @@ rendering. Some of the sections rely on included files, so GitHub rendering is
+TIP: To facilitate documentation edits, you can edit the current page from the `Edit this Page` link located in the upper right corner sidebar. The link opens
+an edit `UI` directly on `GitHub` for the main source file for the current page. These links are
+only present in the `HTML5` version of this reference guide. They look like the following link:
+link:https://github.com/reactor/reactor-core/edit/main/docs/modules/ROOT/pages/aboutDoc.adoc[Edit this Page^, role="fa fa-edit"] to xref:aboutDoc.adoc[About the Documentation].
```
```suggestion
link:https://github.com/reactor/reactor-core/edit/main/docs/modules/ROOT/pages/aboutDoc.adoc[Edit this Page^, role="fa fa-edit"] to make changes to xref:aboutDoc.adoc[About the Documentation] page.
```

pderop replied:
see https://github.com/reactor/reactor-core/pull/3806/commits/75f03a8aba7c45a936b4eefe26ced8fcb250275b

chemicL commented on docs/modules/ROOT/pages/coreFeatures/programmatically-creating-sequence.adoc:
This section renders incorrectly starting from "`onComplete`)"

pderop replied:
fixed in https://github.com/reactor/reactor-core/pull/3806/commits/1ba2fbf6de12ba58e69b1fa3fbe95b6cdfbe8e98

chemicL commented on docs/modules/ROOT/pages/producing.adoc:
```diff
@@ -25,9 +25,10 @@ state, and your generator function now returns a new state on each round.
 For instance, you could use an `int` as the state:
 
 .Example of state-based `generate`
```
The `generate` does not render correctly as code.

pderop replied:
I cannot find a solution for the moment.
what I see is that when also generating the doc into PDF, then the `generate` part is rendered:

![pdf](https://github.com/reactor/reactor-core/assets/650758/b757da22-1b40-4333-802a-3f4b310f9adc)

Now, I have tried to change the backquotes with "#", like this:

```
.Example of state-based #generate#
```

and `generate` is then rendered like this:

![Screenshot 2024-05-15 at 19 20 31](https://github.com/reactor/reactor-core/assets/650758/fa6b905d-3272-40fa-9476-66358cffb019)

are you ok to go with this ? else, I think we need to address this issue later, using another PR ?

chemicL replied:
I don't think the green background works. Let's see if this can be improved afterwards.

pderop replied:
Have created #3813 issue to not forget about this.

chemicL commented on docs/modules/ROOT/pages/kotlin.adoc:
```diff
@@ -82,8 +80,8 @@ declarations and expressive "`value or no value`" semantics without paying the c
 
-Although Java does not let one express null safety in its type-system, Reactor <<null-safety,now
-provides null safety>> of the whole Reactor API through tooling-friendly annotations declared
+Although Java does not let one express null safety in its type-system, Reactor xref:kotlin.adoc#kotlin-null-safety[now provides null safety]
```
The link is incorrect - it should point to the section in advanced features / null-safety, not this section itself.

pderop replied:
fixed in https://github.com/reactor/reactor-core/pull/3806/commits/b80c9d3be77658c74bf28e3e5d1ab05e330394a4

chemicL commented on docs/modules/ROOT/pages/testing.adoc:
```diff
@@ -13,7 +13,6 @@ To use it in your tests, you must add it as a test dependency.
 The following example shows how to add `reactor-test` as a dependency in Maven:
 
 .reactor-test in Maven, in `<dependencies>`
```
I don't know whether or how this can be fixed but please verify ALL the `Example N. Title` places in the documentation - some of them now start with the word "Example", some don't. In the current documentation all examples have the same prefix and a significant and useful number. It would be a good idea to keep that.

chemicL replied:
The other problem is that the rendering of these "Example" headers don't provide a background for the monospace font and it is inconsistent with what we had.

pderop replied:
> I don't know whether or how this can be fixed but please verify ALL the Example N. Title places in the documentation - some of them now start with the word "Example", some don't. In the current documentation all examples have the same prefix and a significant and useful number. It would be a good idea to keep that.

The problem is that before, all  example titles were  indeed **automatically** prefixed with `Example N`, but I cannot find a way to re-enable this behavior.

So, before, we had few titles starting with `Example` word, like these ones:

```
.Example of Callback Hell
.Example of Reactor code equivalent to callback code
.Example of Reactor code with timeout and fallback
.Example of `CompletableFuture` combination
.Example of Reactor code equivalent to future code
.Example of state-based `generate`
```

which were respectively rendered like this:

```
Example 5. Example of Callback Hell
Example 6. Example of Reactor code equivalent to callback code
Example 7. Example of Reactor code with timeout and fallback
Example 8. Example of CompletableFuture combination
Example 9. Example of Reactor code equivalent to future code
Example 11. Example of state-based generate
```

So, now, for these titles, they are rendered without the `Example N`prefix, but we still see the `Example` word, because these few examples title are starting with the `Example` word:

```
Example of Callback Hell
Example of Reactor code equivalent to callback code
Example of Reactor code with timeout and fallback
Example of CompletableFuture combination
Example of Reactor code equivalent to future code
Example of state-based generate
```

> The other problem is that the rendering of these "Example" headers don't provide a background for the monospace font and it is inconsistent with what we had and what other monospaced highlights provide.

I suspect that this is also related to the antora-ui-projectreactor (I'm not sure).

Since for the moment, I'm unable to re-enable automatic generation of the `Example N.` prefix, I would prefer to address this issue in a specific new PR, if you don't mind.

violetagg replied:
If you decide to make a change in additional PR, please create an issue so that we do not forget. Thanks!

pderop replied:
https://github.com/reactor/reactor-core/issues/3811

chemicL commented on docs/modules/ROOT/pages/metrics.adoc:
```diff
@@ -40,11 +40,12 @@ Scheduler schedulerWithMetrics = Micrometer.timedScheduler(
+See xref:metrics-details.adoc#micrometer-details-timedScheduler[Micrometer.timedScheduler()] for produced meters and associated default tags.
 
-// FIXME reactor-monitoring-demo won't be in sync with 3.5.0 anymore
+FIXME reactor-monitoring-demo won't be in sync with 3.5.0 anymore
```
This bit is supposed to be hidden

pderop replied:
Fixed in https://github.com/reactor/reactor-core/pull/3806/commits/b2f925c2dac4632726f5452c615adc2fc108098d

chemicL commented on docs/modules/ROOT/pages/metrics-details.adoc:
```diff
@@ -1,30 +1,28 @@
-:root-target: ./../../build/documentedMetrics/
-
-### Meters and tags for Reactor-Core-Micrometer module
+= Meters and tags for Reactor-Core-Micrometer module
```
This part renders inappropriately as a top-level header inside the body of the metrics.adoc. It should be at the same level as `Scheduler metrics` and `Publisher metrics` headers.

chemicL replied:
This issue is also causing the lower-level headers such as `Micrometer.metrics()` and their descendants to be a level higher than they should. Please consider whether they can be properly adjusted or what are the options if not.

pderop replied:
I think it's fixed in https://github.com/reactor/reactor-core/pull/3806/commits/fad1622c1da5ab72ea90732c09a8d49f5b421489

chemicL replied:
Yes, thank you, now it looks ok.

chemicL commented:
General comment around the change of structure and URIs – do you have a plan to redirect existing links to our documentation out in the wild to the new structure? E.g. https://www.canva.dev/blog/engineering/enabling-real-time-collaboration-with-rsocket/ in the section about Logs has a link to https://projectreactor.io/docs/core/release/reference/#faq.mdc which as I understand would now fail. I believe it would be worth considering a way to make some redirections to aid users in finding this content.

chemicL commented on docs/modules/ROOT/pages/appendices.adoc:
```diff
@@ -0,0 +1,27 @@
+= Appendices
```
This page is quite big, would it make sense to have each appendix as a nested document just like in the core and advanced features ?

pderop replied:
making sense. I did the same that spring-pulsar is doing.
However, we loose automatic appendix labels (A.1, A.2, etc ...).
Actually, to be able to generate appendix labels, we need to include all appendices in one single page ...


anyway, I tend to think that it is now better to have shorter separate pages for the appendices, and spring-pulsar is doing the same: https://docs.spring.io/spring-pulsar/docs/current/reference/appendix/version-compatibility.html

Fixed in https://github.com/reactor/reactor-core/pull/3806/commits/63d66615d18724a709dbc6841b763dbf22fd71b9

chemicL replied:
I think it works much much better now, thanks. I don't think the labeling is as important as the content and ease of access.

chemicL commented on docs/modules/ROOT/pages/apdx-operatorChoice.adoc:
```diff
@@ -1,111 +1,111 @@
 
-* <<which.filtering>>
+* xref:apdx-operatorChoice.adoc#which.filtering[Filtering a Sequence]
 
-* <<which.peeking>>
+* xref:apdx-operatorChoice.adoc#which.peeking[Peeking into a Sequence]
```
I know it was like this before, but these are out of order -> filtering is number 4 and peeking is number 3; can you reorder them here? :)

pderop replied:
this one, I can do it easily :sweat_smile:

fixed in https://github.com/reactor/reactor-core/pull/3806/commits/8ea88631a0d7bf520b77db3f31a5d44d8fab2be3

chemicL commented on docs/modules/ROOT/pages/apdx-operatorChoice.adoc:
```diff
@@ -1,111 +1,111 @@
+* xref:apdx-operatorChoice.adoc#which.time[Working with Time]
 
-* <<which.window>>
+* xref:apdx-operatorChoice.adoc#which.window[Splitting a {javadoc}/reactor/core/publisher/Flux.html[Flux]]
```
I suggest not linking Flux in the link to a header, it's not great UX if you open the javadoc if you point to the Flux word by mistake.

pderop replied:
Fixed in https://github.com/reactor/reactor-core/pull/3806/commits/9243ea5bd2f274cf4e514e064ad88014f0c09afb

chemicL commented on docs/modules/ROOT/pages/apdx-operatorChoice.adoc:
```diff
@@ -1,111 +1,111 @@
+* xref:apdx-operatorChoice.adoc#which.blocking[Going Back to the Synchronous World]
 
-* <<which.multicasting>>
+* xref:apdx-operatorChoice.adoc#which.multicasting[Multicasting a {javadoc}/reactor/core/publisher/Flux.html[Flux] to several https://www.reactive-streams.org/reactive-streams-1.0.3-javadoc/org/reactivestreams/Subscriber.html?is-external=true[Subscribers]]
```
Same comment as above for both links. It's ok if the actual header has these but the link to the header should be simple and be a single atomic link.

pderop replied:
Fixed in https://github.com/reactor/reactor-core/pull/3806/commits/bc98827e48210fa62a1bd0bdd9c44ab96c7d3d41

chemicL commented on docs/modules/ROOT/pages/faq.adoc:
```diff
@@ -1,25 +1,25 @@
-* <<faq.mdc>>
+* xref:faq.adoc#faq.wrap-blocking[How Do I Wrap a Synchronous, Blocking Call?]
+* xref:faq.adoc#faq.chain[I Used an Operator on my `Flux` but it Doesn't Seem to Apply. What Gives?]
+* xref:faq.adoc#faq.monoThen[My `Mono` `zipWith` or `zipWhen` is never called]
+* xref:faq.adoc#faq.retryWhen[How to Use `retryWhen` to Emulate `retry(3)`?]
```
I just noticed the C.4. section has no link to it here (it is also missing in the current documentation), can you please add it?

pderop replied:
Fixed in https://github.com/reactor/reactor-core/pull/3806/commits/539a40486790b75db40d9781ebf99a576c85403a

pderop commented:
with https://github.com/reactor/reactor-core/pull/3806/commits/d3762ab6a722495a575e867e127a0c32ca599ecb, a friendly message is now logged in case 'antora' or 'docs' tasks are used when JDK version is lower than 17.

pderop commented:
> General comment around the change of structure and URIs – do you have a plan to redirect existing links to our documentation out in the wild to the new structure? E.g. https://www.canva.dev/blog/engineering/enabling-real-time-collaboration-with-rsocket/ in the section about Logs has a link to https://projectreactor.io/docs/core/release/reference/#faq.mdc which as I understand would now fail. I believe it would be worth considering a way to make some redirections to aid users in finding this content.

in some cases, we could possibly use `page-aliases` [attributes](https://docs.antora.org/antora/latest/page/page-aliases/). but I'm afraid that for some other cases, it's won't be possible.
https://projectreactor.io/docs/core/release/reference/#faq.mdc will map to the start-page followed by #faq.mdc (https://projectreactor.io/docs/core/release/reference/aboutDoc.html#faq.mdc).

For this particular case, we need to configure some redirect mapping in Cloudflare.

I will create an issue separately to handle these points outside of this PR.

pderop commented:
@chemicL ,

I think I have addressed your feedbacks, except the following:

- `monospace` formatting seems to be not rendered when used in Example titles, I need your feedback about doing an alternative, or we need to address this later in a separate PR, maybe it's related to antora-ui-projectreactor. See https://github.com/reactor/reactor-core/pull/3806#discussion_r1602005716 
- Example titles are not prefixed with ` Example N.` anymore, I have created a separate issue, see https://github.com/reactor/reactor-core/pull/3806#discussion_r1603313977
- for links out in the wild that are pointing to the current documentation, I have created a separate issue, see https://github.com/reactor/reactor-core/pull/3806#issuecomment-2115986506

please have a look to my new commits, when you have time.
thanks!

chemicL commented:
Last remark: `-PforcePdf` does not fail if the `asciidoctor` command was not installed. I think it should fail the build if this option is used. I think this can be addressed in another PR.
